### PR TITLE
Improve cache for staticPrice for products with combinations

### LINKED ISSUE #9803: [BOOM-4368] Multiple SQL calls to the "cart_product" table
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-4368).

- _**Reporter:**_ gato132544
- _**Created at:**_ Thu, 23 Nov 2017 23:51:41 +0100

<p>After much research, a problem has been discovered in the "getPriceStatic" function of the "Product" class.</p>

<p>The so-called SQL is repeated many times in the "cart_product" table, because the "$ id_product_attribute" is not taken into account to be saved as a cache.</p>

<p>Line 2883 to 2893: (Original code)</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
$cache_id = <span class="code-quote">'Product::getPriceStatic_'</span>.(<span class="code-object">int</span>)$id_product.<span class="code-quote">'-'</span>.(<span class="code-object">int</span>)$id_cart;
<span class="code-keyword">if</span> (!Cache::isStored($cache_id) || ($cart_quantity = Cache::retrieve($cache_id) != (<span class="code-object">int</span>)$quantity)) {
$sql = 'SELECT SUM(`quantity`)
FROM `<span class="code-quote">'.DB_PREFIX.'</span>cart_product`
WHERE `id_product` = <span class="code-quote">'.(<span class="code-object">int</span>)$id_product.'</span>
AND `id_cart` = '.(<span class="code-object">int</span>)$id_cart;
$cart_quantity = (<span class="code-object">int</span>)Db::getInstance(PS_USE_SQL_SLAVE)->getValue($sql);
Cache::store($cache_id, $cart_quantity);
} <span class="code-keyword">else</span> {
$cart_quantity = Cache::retrieve($cache_id);
}
</pre>
</div></div>
<p>New code:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
$cache_id = <span class="code-quote">'Product::getPriceStatic_'</span>.(<span class="code-object">int</span>)$id_product.<span class="code-quote">'-'</span>.$id_product_attribute.<span class="code-quote">'-'</span>.(<span class="code-object">int</span>)$id_cart;
<span class="code-keyword">if</span> (!Cache::isStored($cache_id) || ($cart_quantity = Cache::retrieve($cache_id) != (<span class="code-object">int</span>)$quantity)) {
$sql = 'SELECT SUM(`quantity`)
FROM `<span class="code-quote">'._DB_PREFIX_.'</span>cart_product`
WHERE `id_product` = <span class="code-quote">'.(<span class="code-object">int</span>)$id_product.'</span>
AND `id_cart` = <span class="code-quote">'.(<span class="code-object">int</span>)$id_cart.'</span>
AND `id_product_attribute` = '.(<span class="code-object">int</span>)$id_product_attribute;
$cart_quantity = (<span class="code-object">int</span>)Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue($sql);
Cache::store($cache_id, $cart_quantity);
} <span class="code-keyword">else</span> {
$cart_quantity = Cache::retrieve($cache_id);
}
</pre>
</div></div>
<p> </p>
        
* How to reproduce the issue ?

We happened with a customer who had a very large cart, of 77 products for example, where there were only about 20 products, the other 57 were different combinations of the products.
<br/>

<br/>
With the constant &quot;_PS_DEBUG_PROFILING_&quot; actived, under SQL 9200 Called to 2300 generally making a lot of performance.

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Initial PR #8542 from @jbustamanterendon<br>The "id_product_attribute" was not being cached when checking the quantity of each product in the cart, making multiple calls to the "cart_product" table and lowering the store's performance.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #9803 & http://forge.prestashop.com/browse/BOOM-4368
| How to test?  | Many products with many combinations in the cart, for example : 77 products in the shopping cart where there were only about 20 products, the other 57 were different combinations of the products.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16404)

### CODE REVIEW

**Comment by Progi1984** [comment]
This is a remake of this PR: #8542

I fixed the conflicts, and made a new PR because I didn't want to push on @jbustamanterendon's own develop branch, since he made the fix on his develop branch.

File: classes/Product.php [hunk 1]
```diff
@@ -3195,12 +3195,13 @@ public static function getPriceStatic(
...
         if ((int) $id_cart) {
-            $cache_id = 'Product::getPriceStatic_' . (int) $id_product . '-' . (int) $id_cart;
+            $cache_id = 'Product::getPriceStatic_' . (int) $id_product . '-' . (int) $id_product_attribute . '-' . (int) $id_cart;
             if (!Cache::isStored($cache_id) || ($cart_quantity = Cache::retrieve($cache_id) != (int) $quantity)) {
                 $sql = 'SELECT SUM(`quantity`)
                 FROM `' . _DB_PREFIX_ . 'cart_product`
                 WHERE `id_product` = ' . (int) $id_product . '
-                AND `id_cart` = ' . (int) $id_cart;
+                AND `id_cart` = ' . (int) $id_cart . '
+				AND `id_product_attribute` = ' . (int) $id_product_attribute;
```
Comment (PierreRambaud) [inline classes/Product.php]: Watch out, not sure you can request `id_product_attribute` if `$id_product_attribute` is null.
You should add conditionnal `AND`

Comment (Progi1984) [reply classes/Product.php]: Fixed. Thanks.

File: classes/Product.php [hunk 2]
```diff
@@ -3195,12 +3195,13 @@ public static function getPriceStatic(
...
         $cart_quantity = 0;
         if ((int) $id_cart) {
-            $cache_id = 'Product::getPriceStatic_' . (int) $id_product . '-' . (int) $id_cart;
+            $cache_id = 'Product::getPriceStatic_' . (int) $id_product . '-' . (int) $id_product_attribute . '-' . (int) $id_cart;
             if (!Cache::isStored($cache_id) || ($cart_quantity = Cache::retrieve($cache_id) != (int) $quantity)) {
                 $sql = 'SELECT SUM(`quantity`)
                 FROM `' . _DB_PREFIX_ . 'cart_product`
                 WHERE `id_product` = ' . (int) $id_product . '
                 AND `id_cart` = ' . (int) $id_cart;
+                $sql .= !empty($id_product_attribute) ? ' AND `id_product_attribute` = ' . (int) $id_product_attribute : '';
```
Comment (PierreRambaud) [inline classes/Product.php]:
```suggestion
                $sql .= $id_product_attribute !== null ? ' AND `id_product_attribute` = ' . (int) $id_product_attribute : '';
```

**Review by PierreRambaud — APPROVED** [review]

**Comment by matks** [comment]
Thank you @Progi1984 and @jbustamanterendon

**Pull request merged**